### ISSUE TITLE
Failed to automatically apply fixes

### ISSUE THREAD

**Issue by Tiwalun** [rust-lang/rust#107480]
I tried running

```bash
cargo clippy --fix
```

I expected to see this happen: Some fixes get applied, or not applied.

Instead, this happened: 

Some fixes could not be applied, and rustc / cargo asked me to file an issue here:

<details><summary>cargo clippy --fix output</summary>
<p>

```
 cargo clippy --fix
   Compiling probe-rs-target v0.16.0 (/Users/tiwalun/code/probe-rs/probe-rs-target)
    Checking xtask v0.16.0 (/Users/tiwalun/code/probe-rs/xtask)
   Compiling probe-rs v0.16.0 (/Users/tiwalun/code/probe-rs/probe-rs)
warning: failed to automatically apply fixes suggested by rustc to crate `probe_rs`

after fixes were automatically applied the compiler reported errors within these files:

  * probe-rs/src/probe/stlink/mod.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0425]: cannot find value `value` in this scope
   --> probe-rs/src/probe/stlink/mod.rs:886:12
    |
886 |         Ok(value)
    |            ^^^^^ not found in this scope

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
Original diagnostics will follow.

warning: useless conversion to the same type: `probe::stlink::StlinkError`
   --> probe-rs/src/probe/stlink/usb_interface.rs:235:24
    |
235 |               return Err(StlinkError::NotEnoughBytesWritten {
    |  ________________________^
236 | |                 is: written_bytes,
237 | |                 should: CMD_LEN,
238 | |             }
239 | |             .into());
    | |___________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
    = note: `#[warn(clippy::useless_conversion)]` on by default
help: consider removing `.into()`
    |
235 ~             return Err(StlinkError::NotEnoughBytesWritten {
236 +                 is: written_bytes,
237 +                 should: CMD_LEN,
238 ~             });
    |

warning: useless conversion to the same type: `probe::stlink::StlinkError`
   --> probe-rs/src/probe/stlink/mod.rs:300:25
    |
300 |                     Err(StlinkError::VoltageDivisionByZero.into())
    |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `StlinkError::VoltageDivisionByZero`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: this let-binding has unit value
   --> probe-rs/src/probe/stlink/mod.rs:863:9
    |
863 | /         let value = retry_on_wait(|| {
864 | |             self.device.write(
865 | |                 &[
866 | |                     commands::JTAG_COMMAND,
...   |
881 | |             self.get_last_rw_status()
882 | |         })?;
    | |____________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_unit_value
    = note: `#[warn(clippy::let_unit_value)]` on by default
help: omit the `let` binding
    |
863 ~         retry_on_wait(|| {
864 +             self.device.write(
865 +                 &[
866 +                     commands::JTAG_COMMAND,
867 +                     commands::JTAG_READMEM_32BIT,
868 +                     addbytes[0],
869 +                     addbytes[1],
870 +                     addbytes[2],
871 +                     addbytes[3],
872 +                     data_length[0],
873 +                     data_length[1],
874 +                     apsel,
875 +                 ],
876 +                 &[],
877 +                 data,
878 +                 TIMEOUT,
879 +             )?;
880 +
881 +             self.get_last_rw_status()
882 +         })?;
    |

warning: useless conversion to the same type: `probe::stlink::StlinkError`
    --> probe-rs/src/probe/stlink/mod.rs:1841:14
     |
1841 |             &StlinkError::BanksNotAllowedOnDPRegister.into()
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `StlinkError::BanksNotAllowedOnDPRegister`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `probe::stlink::StlinkError`
    --> probe-rs/src/probe/stlink/mod.rs:1844:14
     |
1844 |             &StlinkError::CommandFailed(Status::JtagFreqNotSupported).into()
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `StlinkError::CommandFailed(Status::JtagFreqNotSupported)`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `probe::stlink::StlinkError`
    --> probe-rs/src/probe/stlink/mod.rs:1847:14
     |
1847 |             &StlinkError::CommandFailed(Status::SwdDpWait).into()
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `StlinkError::CommandFailed(Status::SwdDpWait)`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `probe::stlink::StlinkError`
    --> probe-rs/src/probe/stlink/mod.rs:1850:14
     |
1850 |             &StlinkError::CommandFailed(Status::SwdApWait).into()
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `StlinkError::CommandFailed(Status::SwdApWait)`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: `probe-rs` (lib test) generated 7 warnings (3 duplicates)
warning: `probe-rs` (lib) generated 3 warnings
    Checking probe-rs-cli-util v0.16.0 (/Users/tiwalun/code/probe-rs/probe-rs-cli-util)
    Checking gdb-server v0.16.0 (/Users/tiwalun/code/probe-rs/gdb-server)
    Checking target-gen v0.16.0 (/Users/tiwalun/code/probe-rs/target-gen)
    Checking rtthost v0.16.0 (/Users/tiwalun/code/probe-rs/rtthost)
    Checking smoke_tester v0.16.0 (/Users/tiwalun/code/probe-rs/smoke-tester)
    Checking probe-rs-debugger v0.16.0 (/Users/tiwalun/code/probe-rs/debugger)
warning: failed to automatically apply fixes suggested by rustc to crate `probe_rs`

after fixes were automatically applied the compiler reported errors within these files:

  * probe-rs/src/probe/stlink/mod.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0425]: cannot find value `value` in this scope
   --> probe-rs/src/probe/stlink/mod.rs:886:12
    |
886 |         Ok(value)
    |            ^^^^^ not found in this scope

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
Original diagnostics will follow.

warning: `probe-rs` (lib) generated 3 warnings (3 duplicates)
   Compiling cargo-flash v0.16.0 (/Users/tiwalun/code/probe-rs/cargo-flash)
   Compiling cargo-embed v0.16.0 (/Users/tiwalun/code/probe-rs/cargo-embed)
   Compiling probe-rs-cli v0.16.0 (/Users/tiwalun/code/probe-rs/cli)
    Finished dev [unoptimized + debuginfo] target(s) in 1m 02s
```
</p>
</details>

The error occured while trying to compile this commit: https://github.com/probe-rs/probe-rs/commit/3a192909b544c03e1eb3b5da8ea0476f9d75de9d

### Meta

`rustc --version --verbose`:
```
rustc 1.67.0 (fc594f156 2023-01-24)
binary: rustc
commit-hash: fc594f15669680fa70d255faec3ca3fb507c3405
commit-date: 2023-01-24
host: aarch64-apple-darwin
release: 1.67.0
LLVM version: 15.0.6
```

The error also occurs with the following nightly version:
```
rustc 1.69.0-nightly (e972bc808 2023-01-29)
binary: rustc
commit-hash: e972bc8083d5228536dfd42913c8778b6bb04c8e
commit-date: 2023-01-29
host: aarch64-apple-darwin
release: 1.69.0-nightly
LLVM version: 15.0.7
```

**Comment by Tiwalun** [comment]
Smaller repro:

```rust
fn main() {
    test().unwrap();
}

fn test() -> Result<(), String> {
    let unit_binding = ();

    Ok(unit_binding)
}
```

This causes the same error, when running `cargo clippy --fix`.

**Comment by lukas-code** [comment]
Probably duplicate of https://github.com/rust-lang/rust-clippy/issues/9893
